unit test fails

go version go1.9.7 linux/amd64

git commit hash - 83672023c79e6b352bf9d4908e6c060c7a24323c


```
[vodolaz095@tractor go-dkim]$ go test -v
=== RUN   Test_NewSigOptions
--- PASS: Test_NewSigOptions (0.00s)
=== RUN   Test_SignConfig
--- PASS: Test_SignConfig (0.01s)
=== RUN   Test_canonicalize
--- PASS: Test_canonicalize (0.00s)
=== RUN   Test_Sign
--- PASS: Test_Sign (0.01s)
=== RUN   Test_Verify
--- FAIL: Test_Verify (0.47s)
	dkim_test.go:350: 
			Error Trace:	dkim_test.go:350
			Error:      	Not equal: 
			            	expected: &errors.errorString{s:"body hash did not verify"}
			            	actual  : &errors.errorString{s:"no key for verify"}
			            	
			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -1,2 +1,2 @@
			            	-(*errors.errorString)(body hash did not verify)
			            	+(*errors.errorString)(no key for verify)
			            	 
			Test:       	Test_Verify
	dkim_test.go:351: 
			Error Trace:	dkim_test.go:351
			Error:      	Not equal: 
			            	expected: 6
			            	actual  : 2
			Test:       	Test_Verify
	dkim_test.go:375: 
			Error Trace:	dkim_test.go:375
			Error:      	Received unexpected error:
			            	no key for verify
			Test:       	Test_Verify
	dkim_test.go:376: 
			Error Trace:	dkim_test.go:376
			Error:      	Not equal: 
			            	expected: 1
			            	actual  : 2
			Test:       	Test_Verify
	dkim_test.go:381: 
			Error Trace:	dkim_test.go:381
			Error:      	Received unexpected error:
			            	no key for verify
			Test:       	Test_Verify
	dkim_test.go:382: 
			Error Trace:	dkim_test.go:382
			Error:      	Not equal: 
			            	expected: 1
			            	actual  : 2
			Test:       	Test_Verify
FAIL
exit status 1
FAIL	github.com/toorop/go-dkim	0.492s
```

--- a/dkim.go
+++ b/dkim.go
@@ -200,7 +200,7 @@ func Sign(email *[]byte, options SigOptions) error {
 // state: SUCCESS or PERMFAIL or TEMPFAIL, TESTINGSUCCESS, TESTINGPERMFAIL
 // TESTINGTEMPFAIL or NOTSIGNED
 // error: if an error occurs during verification
-func Verify(email *[]byte) (verifyOutput, error) {
+func Verify(email *[]byte, opts ...DNSOpt) (verifyOutput, error) {
 	// parse email
 	dkimHeader, err := newDkimHeaderFromEmail(email)
 	if err != nil {
@@ -211,7 +211,7 @@ func Verify(email *[]byte) (verifyOutput, error) {
 	}
 
 	// we do not set query method because if it's others, validation failed earlier
-	pubKey, verifyOutputOnError, err := NewPubKeyRespFromDNS(dkimHeader.Selector, dkimHeader.Domain)
+	pubKey, verifyOutputOnError, err := NewPubKeyRespFromDNS(dkimHeader.Selector, dkimHeader.Domain, opts...)
 	if err != nil {
 		// fix https://github.com/toorop/go-dkim/issues/1
 		//return getVerifyOutput(verifyOutputOnError, err, pubKey.FlagTesting)

--- a/dkim_test.go
+++ b/dkim_test.go
@@ -5,6 +5,7 @@ import (
 	"crypto/rsa"
 	"crypto/x509"
 	"encoding/pem"
+	"net"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -353,53 +354,69 @@ func Test_Sign(t *testing.T) {
 }
 
 func Test_Verify(t *testing.T) {
+	resolveTXT := DNSOptLookupTXT(func(name string) ([]string, error) {
+		switch name {
+		case selector + "._domainkey." + domain:
+			return []string{"v=DKIM1; t=y; p=" + pubKey}, nil
+		// case "TODO._domainkey.gmail.com":
+		// 	return []string{"v=DKIM1; p="}, nil
+		default:
+			return net.LookupTXT(name)
+		}
+	})
+
 	// no DKIM header
 	email := []byte(emailBase)
-	status, err := Verify(&email)
+	status, err := Verify(&email, resolveTXT)
 	assert.Equal(t, NOTSIGNED, status)
 	assert.Equal(t, ErrDkimHeaderNotFound, err)
 
 	// No From
 	email = []byte(signedNoFrom)
-	status, err = Verify(&email)
+	status, err = Verify(&email, resolveTXT)
 	assert.Equal(t, ErrVerifyBodyHash, err)
 	assert.Equal(t, TESTINGPERMFAIL, status) // cause we use dkheader of the "with from" email
 
 	// missing mandatory 'a' flag
 	email = []byte(signedMissingFlag)
-	status, err = Verify(&email)
+	status, err = Verify(&email, resolveTXT)
 	assert.Error(t, err)
 	assert.Equal(t, PERMFAIL, status)
 	assert.Equal(t, ErrDkimHeaderMissingRequiredTag, err)
 
 	// missing bad algo
 	email = []byte(signedBadAlgo)
-	status, err = Verify(&email)
+	status, err = Verify(&email, resolveTXT)
 	assert.Equal(t, PERMFAIL, status)
 	assert.Equal(t, ErrSignBadAlgo, err)
 
 	// bad a flag
 	email = []byte(signedBadAFlag)
-	status, err = Verify(&email)
+	status, err = Verify(&email, resolveTXT)
 	assert.Equal(t, PERMFAIL, status)
 	assert.Equal(t, ErrSignBadAlgo, err)
 
 	// relaxed
 	email = []byte(signedRelaxedRelaxedLength)
-	status, err = Verify(&email)
+	status, err = Verify(&email, resolveTXT)
 	assert.NoError(t, err)
 	assert.Equal(t, SUCCESS, status)
 
 	// simple
 	email = []byte(signedSimpleSimpleLength)
-	status, err = Verify(&email)
+	status, err = Verify(&email, resolveTXT)
 	assert.NoError(t, err)
 	assert.Equal(t, SUCCESS, status)
 
 	// gmail
-	email = []byte(fromGmail)
-	status, err = Verify(&email)
-	assert.NoError(t, err)
-	assert.Equal(t, SUCCESS, status)
+	// TODO:
+	// Google removed this DNS record some time ago. Someone will have to send an email they're
+	// OK with being publicly available, replace the value of the fromGmail var with that, then grab
+	// the DNS record indicated in the DKIM signature and update the resolveTXT function to return
+	// it when asked. Then this should work.
+	// email = []byte(fromGmail)
+	// status, err = Verify(&email, resolveTXT)
+	// assert.NoError(t, err)
+	// assert.Equal(t, SUCCESS, status)
 
 }

--- a/pubKeyRep.go
+++ b/pubKeyRep.go
@@ -22,10 +22,45 @@ type PubKeyRep struct {
 	FlagIMustBeD bool // flag i
 }
 
+// DNSOptions holds settings for looking up DNS records
+type DNSOptions struct {
+	netLookupTXT func(name string) ([]string, error)
+}
+
+// DNSOpt represents an optional setting for looking up DNS records
+type DNSOpt interface {
+	apply(*DNSOptions)
+}
+
+type dnsOpt func(*DNSOptions)
+
+func (opt dnsOpt) apply(dnsOpts *DNSOptions) {
+	opt(dnsOpts)
+}
+
+// DNSOptLookupTXT sets the function to use to lookup TXT records.
+//
+// This should probably only be used in tests.
+func DNSOptLookupTXT(netLookupTXT func(name string) ([]string, error)) DNSOpt {
+	return dnsOpt(func(opts *DNSOptions) {
+		opts.netLookupTXT = netLookupTXT
+	})
+}
+
 // NewPubKeyRespFromDNS retrieves the TXT record from DNS based on the specified domain and selector
 // and parses it.
-func NewPubKeyRespFromDNS(selector, domain string) (*PubKeyRep, verifyOutput, error) {
-	txt, err := net.LookupTXT(selector + "._domainkey." + domain)
+func NewPubKeyRespFromDNS(selector, domain string, opts ...DNSOpt) (*PubKeyRep, verifyOutput, error) {
+	dnsOpts := DNSOptions{}
+
+	for _, opt := range opts {
+		opt.apply(&dnsOpts)
+	}
+
+	if dnsOpts.netLookupTXT == nil {
+		dnsOpts.netLookupTXT = net.LookupTXT
+	}
+
+	txt, err := dnsOpts.netLookupTXT(selector + "._domainkey." + domain)
 	if err != nil {
 		if strings.HasSuffix(err.Error(), "no such host") {
 			return nil, PERMFAIL, ErrVerifyNoKeyForSignature

